Added additional emergent constraints on ECS

schlunma opened the pull request:
Add several emergent constraints (including lots of code for general emergent constraints) for ECS, among them the Volodin (2008) constraint necessary for Lauer et al.

Depends on ESMValGroup/ESMValCore#546 and ESMValGroup/ESMValCore#562.

* * *

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValTool/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] Give this pull request a descriptive title that can be used as a one line summary in a changelog
-   [x] Make sure your code is composed of functions of no more than 50 lines and uses meaningful names for variables
-   [ ] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Preferably Codacy code quality checks pass, however a few remaining hard to solve Codacy issues are still acceptable. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes 
-   [x] (Only if really necessary) Add any additional dependencies needed for the diagnostic script to setup.py, esmvaltool/install/R/r_requirements.txt or esmvaltool/install/Julia/julia_requirements.txt (depending on the language of your script) and also to meta.yaml for conda dependencies (includes Python and others, but not R/Julia)
-   [x] If new dependencies are introduced, check that the license is compatible with [Apache2.0](https://github.com/ESMValGroup/ESMValTool/blob/master/LICENSE)

New recipe/diagnostic

-   [x] Add documentation for the recipe to the `doc/sphinx/source/recipes` folder and add a new entry to index.rst
-   [x] Add provenance information

* * *

Closes #1584.

axel-lauer commented:
Looks already pretty good to me. Would it be (theoretically speaking) be possible to use the newly defined functions "sherwood_d" and "sherwood_s" in the calculation of ltmi (ecs_scatter.ncl). Would be nice for consistency reasons but might not be worth the effort.

schlunma commented:
I think so, but since I'm not really fluent in NCL I was scared of braking something. I will give it a try though!

schlunma commented:
This is ready for review now. Please note that you need a file called `ecs_emergent_constraints.yml` in your `auxiliary_data_dir` with the following content:

```yaml
# Prediction input
- dataset: 0
  project: OBS
  var_type: prediction_input
  tag: BRS
  data: 44.5
- dataset: 0
  project: OBS
  var_type: prediction_input
  tag: VOL
  data: -25
- dataset: 0
  project: OBS
  var_type: prediction_input
  tag: ZHA
  data: -1.28


# Prediction input error
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: SU
  data: 0.25
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: BRS
  data: 3.5
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: VOL
  data: 5.5
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: ZHA
  data: 0.187
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: COX
  data: 0.05
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: SHL
  data: 0.08
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: SHD
  data: 0.031
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: SHS
  data: 0.05
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: BRA
  data: 0.13
- dataset: 0
  project: OBS
  var_type: prediction_input_error
  tag: LIP
  data: 1.75
- dataset: 0
  project: obs4mips
  var_type: prediction_input_error
  tag: TII
  data: 0.5
- dataset: 0
  project: obs4mips
  var_type: prediction_input_error
  tag: TIH
  data: 10.0
```

schlunma commented:
I have no idea why the test is failing:

```
ImportError: cannot import name 'add_plev_from_altitude' from 'esmvalcore.cmor._fixes.shared'
```

 I had this issue earlier because of a circular dependency but I though this was fixed by ESMValGroup/ESMValCore#604 with the introduction of the module ` esmvalcore/iris_helpers.py`.

Locally the diagnostic works perfectly fine. Additionally, using `from esmvalcore.cmor._fixes.shared import add_plev_from_altitude` in the python command line works just fine.

axel-lauer commented on esmvaltool/recipes/recipe_ecs_constraints.yml:
```diff
@@ -0,0 +1,1131 @@
+# recipe_ecs_multivariate_constraint.yml
```
Should be recipe_ecs_constraint.yml

bouweandela commented on esmvaltool/diag_scripts/emergent_constraints/ecs_scatter.py:
```diff
@@ -0,0 +1,813 @@
+
+import esmvaltool.diag_scripts.emergent_constraints as ec
+import esmvaltool.diag_scripts.shared.iris_helpers as ih
+from esmvalcore.cmor._fixes.shared import (add_plev_from_altitude,
+                                           add_sigma_factory)
```
Please only import things from ESMValCore that are part of the public API, documented here: https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.html
Using undocumented and private (i.e. modules/functions starting with an underscore) is not a good idea, because these can be changed without notice in a minor revision.

There are two possible ways to solve this:
1) make the functions you need part of the public API
2) copy paste the required code here

schlunma replied:
Opened ESMValGroup/ESMValCore#714 to add these functions to the public API.

bouweandela replied:
Note that I changed the module path a bit, the functions are now available in [`esmvalcore.cmor.fixes`.](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/api/esmvalcore.cmor.html#module-esmvalcore.cmor.fixes)

bouweandela commented on esmvaltool/recipes/recipe_ecs_constraints.yml:
```diff
@@ -0,0 +1,1131 @@
+          tag: ECS
+          plot_ylim: [1.5, 6.0]
+          project: CMIP5
+          provenance_authors: ['schlund_manuel']
+          provenance_domains: ['global']
+          provenance_realms: ['atmos']
+          provenance_references: ['gregory04grl']
+          provenance_statistics: ['mean', 'anomaly']
+          provenance_themes: ['phys']
```
`themes` and `realms` is usually included directly under the diagnostic in the recipe, have a look at how it's done in e.g. recipe_perfmetrics_CMIP5.yml or examples/recipe_python.yml. The other attributes are usually set in the code, because they are very closely tied to the diagnostic script.

schlunma replied:
For the diagnostic [univariate_constraint.py](https://github.com/ESMValGroup/ESMValTool/blob/add_ecs_constraints/esmvaltool/diag_scripts/emergent_constraints/univariate_constraint.py) it does not make sense to hardcode the provenance information in the header of the recipe or the diagnostic script itself, since it is able to handle arbitrary emergent constraints. E.g., `realms` could be anything (atmosphere, ocean, land, etc.).

bouweandela replied:
But you will still be the author of the diagnostic script, right?

bouweandela commented:
> I have no idea why the test is failing:
>
> `ImportError: cannot import name 'add_plev_from_altitude' from 'esmvalcore.cmor._fixes.shared'`

This is because the tests in the ESMValTool repository are run against the latest release of ESMValCore. These functions are not in a released version yet. Please also see my comment https://github.com/ESMValGroup/ESMValTool/pull/1585#discussion_r452880962 on importing from private modules above.

schlunma commented:
@axel-lauer I have added the contents of the external file to the recipe, so the external file is not necessary anymore. I also addressed the other comments by @bouweandela. Please note that this PR needs a new release of ESMValCore to work.

mattiarighi commented:
I get an error in `recipe_ecs_constraints.yml`
```
Traceback (most recent call last):
  File "/work/bd0080/b309057/SOFTWARE/miniconda3/envs/esmvaltool/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/work/bd0080/b309057/SOFTWARE/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/esmvalcore/_task.py", line 730, in _run_task
    output_files = task.run()
  File "/work/bd0080/b309057/SOFTWARE/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/esmvalcore/_task.py", line 242, in run
    self.output_files = self._run(input_files)
  File "/work/bd0080/b309057/SOFTWARE/miniconda3/envs/esmvaltool/lib/python3.8/site-packages/esmvalcore/_task.py", line 523, in _run
    raise DiagnosticError(
esmvalcore._task.DiagnosticError: Diagnostic script emergent_constraints/ecs_scatter.py failed with return code 1. See the log in /mnt/lustre02/work/bd0854/b309057/ESMValTool/v2_output/recipe_ecs_constraints_20200721_092906/run/diag_x_volodin_cmip5/ecs_predictor/log.txt
```

In the log:
```
Traceback (most recent call last):
  File "/mnt/lustre01/pf/b/b309057/ESMValTool/public/esmvaltool/diag_scripts/emergent_constraints/ecs_scatter.py", line 59, in <module>
    from esmvalcore.cmor import add_plev_from_altitude, add_sigma_factory
ImportError: cannot import name 'add_plev_from_altitude' from 'esmvalcore.cmor' (/work/bd0080/b309057/SOFTWARE/miniconda3/envs/esmvaltool/lib/python3.8/site-
packages/esmvalcore/cmor/__init__.py)
```

bouweandela commented:
Yes, see https://github.com/ESMValGroup/ESMValTool/pull/1585#discussion_r457625565

mattiarighi commented:
@schlunma can you fix this please?

mattiarighi commented:
I tested successfully both recipes.
@bouweandela if you are ok with this please merge and cherry pick to `v2.0.0-master`.

bouweandela commented:
The way the provenance and tags are implemented in these diagnostics is different from how it's normally done and I'm not convinced it's the right way to go, see also https://github.com/ESMValGroup/ESMValTool/pull/1585#discussion_r452883942. @mattiarighi Do you want to fix this now, or shall we make an issue so @schlunma can fix it when he's back from holiday?

mattiarighi commented:
The problem is that this needs to go in the final release because it's part of @axel-lauer's paper.

What about merging it now and opening an issue that @schlunma can take care of when he is back?

mattiarighi commented:
@bouweandela please cherry-pick in `v2.0.0-release`.

bouweandela commented:
Done!
